src/dataflow/src/source/s3.rs

```diff
@@ -515,27 +552,57 @@ enum Sent {
     SenderClosed,
 }
 
+#[derive(Debug)]
+enum S3Error {
+    BodyMissing(String),
+    ClientConstructionFailed(anyhow::Error),
+    Decode(String, std::io::Error),
+    GetObjectError(RusotoError<rusoto_s3::GetObjectError>),
+    ListObjectsFailed(RusotoError<rusoto_s3::ListObjectsV2Error>),
+    Read(std::io::Error),
+    RetryFailed,
+}
+
+impl std::fmt::Display for S3Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            S3Error::BodyMissing(body) => write!(f, "Get object response had no body: {}", body),
+            S3Error::ClientConstructionFailed(err) => err.fmt(f),
+            S3Error::Decode(key, err) => {
+                write!(f, "Failed to decode object {} using gzip: {}", key, err)
+            }
+            S3Error::GetObjectError(err) => err.fmt(f),
+            S3Error::ListObjectsFailed(err) => err.fmt(f),
+            S3Error::Read(err) => err.fmt(f),
+            S3Error::RetryFailed => write!(f, "Retry failed to produce result"),
+        }
+    }
+}
+
+type S3Result<R> = Result<R, S3Error>;
+
 async fn download_object(
-    tx: &SyncSender<anyhow::Result<InternalMessage>>,
+    tx: &SyncSender<S3Result<InternalMessage>>,
     activator: &SyncActivator,
     client: &S3Client,
-    bucket: String,
-    key: String,
+    bucket: &str,
+    key: &str,
     compression: Compression,
-) -> Option<DownloadMetricUpdate> {
+) -> (bool, Option<DownloadMetricUpdate>) {
     let obj = match client
         .get_object(GetObjectRequest {
-            bucket: bucket.clone(),
+            bucket: bucket.into(),
             key: key.to_string(),
             ..Default::default()
         })
         .await
     {
         Ok(obj) => obj,
         Err(e) => {
-            tx.send(Err(anyhow!("Unable to GET object: {}", e)))
+            tx.send(Err(S3Error::GetObjectError(e)))
                 .unwrap_or_else(|e| log::debug!("unable to send error on stream: {}", e));
-            return None;
+            // TODO(mh): Determine which Rusoto errors can be retried.
```

Comment:
We shouldn't spend *too* much time trying to figure that out because the new official AWS rust client will automatically retry retryable errors. It is unclear how long it will take us to transition over, though, so figuring it out is valuable.